File: study_ws/src/py_pkg/py_pkg/activity004_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from irobot_interfaces.srv import SetLed

# Python Specifics..
import random
from functools import partial


class BatteryNode(Node): 
    def __init__(self):
        super().__init__("battery_node")
        self.get_logger().info('Client has initiated...')
        self.client_ = self.create_client(SetLed, 'set_led')
        self.request_timer = self.create_timer(1.0, self.request_)

    def battery_percent(self):
        return random.randint(1, 100)

    def process_led_states(self, array, charge):
        match charge:
            case x if x < 25:
                array = [False] * len(array)
            case x if x < 50:
                array[0] = True
                array[1:] = [False] * (len(array) - 1)
            case x if x < 75:
                array[0] = True
                array[1] = True
                array[2:] = [False] * (len(array) - 2)
            case x if 75 < x < 95:
                array[0] = True
                array[1] = True
                array[2] = True
                array[3:] = [False] * (len(array) - 3)
            case x if x >= 95:
                array = [True] * len(array)
        return array

    def request_(self):
        request_call = SetLed.Request()
        self.charge = self.battery_percent()
        request_call.input_array = [False] * len(request_call.input_array)
        request_call.input_array = self.process_led_states(request_call.input_array, self.charge)

        self.get_logger().info(f'Determined LED States: {request_call.input_array}')
        if not self.client_.wait_for_service(timeout_sec=2.0):
            self.get_logger().warn('Waiting for Server to start....')
        self.future = self.client_.call_async(request_call)
        self.future.add_done_callback(partial(self.response_callback, request_call))

    def response_callback(self, request, future):
        green = '\033[92m'
        endc = '\033[0m'
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'{green}LED Array Updated | Battery Percent: {self.charge}%{endc}')
            else:
                self.get_logger().warn('LED Array Update Failed.')
        except Exception as e:
            self.get_logger().error(f"Service call failed: {e}")

def main(args=None):
    rclpy.init(args=args)
    node = BatteryNode() 
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: study_ws/src/cpp_pkg/src/activity004_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "irobot_interfaces/srv/set_led.hpp"
#include "irobot_interfaces/msg/led_states.hpp"
#include <cstdlib>
#include <array>

using ServiceResponseFuture = rclcpp::Client<irobot_interfaces::srv::SetLed>::SharedFuture;

class BatteryNode : public rclcpp::Node 
{
public:
    BatteryNode() : Node("battery_node")
    {
        srand(time(0));
        RCLCPP_INFO(this->get_logger(), "Battery Node has initiated...");
        client_ = this->create_client<irobot_interfaces::srv::SetLed>("set_led"); 
        loop_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&BatteryNode::timer_callback, this));
    }

    ~BatteryNode() {
        for (auto &thread : threads_) {
            if (thread.joinable()) {
                thread.join();  // Ensure all threads are joined before destruction
            }
        }
    }


private:
    rclcpp::Client<irobot_interfaces::srv::SetLed>::SharedPtr client_;
    std::vector<std::thread> threads_;
    rclcpp::TimerBase::SharedPtr loop_;
    std::string green = "\033[92m";
    std::string endc = "\033[0m";
    short charge = 0;

    void timer_callback(){
        if (!client_->service_is_ready()) {
            RCLCPP_WARN(this->get_logger(), "Service 'set_led' not available yet. Retrying...");
        }
        else {threads_.emplace_back(std::thread(&BatteryNode::handler, this));}
    }

    void handler(){response(client_request());}

    ServiceResponseFuture client_request(){

        auto request = std::make_shared<irobot_interfaces::srv::SetLed::Request>();
        charge = BatteryLevel();
        request->input_array = process_led_states(request->input_array, charge); // Fixed for std::array<bool, 4>
        std::stringstream ss;
        for (const auto& val : request->input_array) ss << val << " ";
        RCLCPP_INFO(this->get_logger(), "Determined LED States: %s", ss.str().c_str());
        try{
            auto future = client_->async_send_request(request).share(); // Updated: Use .share()
            return future;
        }
        catch(const std::exception &e){
            RCLCPP_ERROR(this->get_logger(), "Error occurred during service call: %s", e.what());
            return ServiceResponseFuture(); // Fixed: Return default-constructed future
        }
    }

    void response(ServiceResponseFuture future){
        if (!future.valid()) {
            RCLCPP_ERROR(this->get_logger(), "Invalid future object.");
            return;
        }
        try {
            auto response = future.get();
            if (response->success) {
                RCLCPP_INFO(this->get_logger(), "%sLED Array Updated | Battery Percent: %d%%%s", green.c_str(), charge, endc.c_str());
            } else {
                RCLCPP_ERROR(this->get_logger(), "LED States Update Failed");
            }
        } catch (const std::exception &e) {
            RCLCPP_ERROR(this->get_logger(), "Error occurred: %s", e.what());
        }
    }

    short int BatteryLevel(){return rand() % 100 + 1;}

    std::array<bool, 4> process_led_states(std::array<bool, 4>& array, short int charge){ // Fixed for std::array
        if (charge < 25) {
            std::fill(array.begin(), array.end(), false);
        } 
        else if (charge < 50) {
            array[0] = true;
            std::fill(array.begin() + 1, array.end(), false);
        } 
        else if (charge < 75) {
            array[0] = true;
            array[1] = true;
            std::fill(array.begin() + 2, array.end(), false);
        } 
        else if (charge > 75 && charge < 95) {
            array[0] = true;
            array[1] = true;
            array[2] = true;
            std::fill(array.begin() + 3, array.end(), false);
        } 
        else if (charge >= 95) {
            std::fill(array.begin(), array.end(), true);
        }

        return array;
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<BatteryNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
